### PR TITLE
ITE: drivers/i2c: I2C driver divided into two compatibles

### PR DESCRIPTION
**Pull request by GTLin08**
As mentioned in #42882, the I2C of IT8XXX2 is designed for two different
IP blocks, so this PR divides this I2C driver into two compatibles.

Signed-off-by: Tim Lin <tim2.lin@ite.corp-partner.google.com>

### CODE REVIEW

File: drivers/i2c/i2c_ite_enhance.c [hunk 1]
```diff
@@ -0,0 +1,776 @@
...
+	const struct i2c_enhance_config *config = dev->config;
+	uint8_t *base = config->base;
+
+	/* State reset and hardware reset */
+	IT8XXX2_I2C_CTR(base) = E_STS_AND_HW_RST;
+}
+
+/* Set clock frequency for i2c port D, E , or F */
+static void i2c_enhanced_port_set_frequency(const struct device *dev,
+					    int freq_khz)
```
Comment (keith-zephyr) [inline drivers/i2c/i2c_ite_enhance.c:211]: This parameter is now the frequency in Hz, not kHz.

File: drivers/i2c/i2c_ite_enhance.c [hunk 2]
```diff
@@ -0,0 +1,776 @@
...
+		 * bus error, device error).
+		 */
+		if (data->err)
+			break;
+
+		if (res != 0) {
+			data->err = ETIMEDOUT;
+			/* reset i2c port */
+			i2c_reset(dev);
+			printk("I2C ch%d:0x%X reset cause %d\n",
```
Comment (keith-zephyr) [inline drivers/i2c/i2c_ite_enhance.c:580]: Use logging instead of printk.

File: drivers/i2c/i2c_ite_enhance.c [hunk 3]
```diff
@@ -0,0 +1,776 @@
...
+
+	volatile uint8_t *reg = (volatile uint8_t *)
+				(IT8XXX2_ECPM_BASE + (offset >> 8));
+	uint8_t reg_mask = offset & 0xff;
+	*reg &= ~reg_mask;
+
+	/* Enable I2C function */
+	/* Software reset */
+	IT8XXX2_I2C_DHTR(base) |= IT8XXX2_I2C_SOFT_RST;
+	IT8XXX2_I2C_DHTR(base) &= 0x7F;
```
Comment (keith-zephyr) [inline drivers/i2c/i2c_ite_enhance.c:643]: Use `~IT8XXX2_I2C_SOFT_RST` instead of 0x7F.

File: drivers/i2c/i2c_ite_enhance.c [hunk 4]
```diff
@@ -0,0 +1,776 @@
...
+	k_mutex_init(&data->mutex);
+	k_sem_init(&data->device_sync_sem, 0, K_SEM_MAX_LIMIT);
+
+	/* Enable clock to specified peripheral */
+	switch ((uint32_t)base) {
+	case DT_REG_ADDR(DT_NODELABEL(i2c3)):
+		offset = CGC_OFFSET_SMBD;
+		break;
+	case DT_REG_ADDR(DT_NODELABEL(i2c4)):
+		offset = CGC_OFFSET_SMBE;
```
Comment (keith-zephyr) [inline drivers/i2c/i2c_ite_enhance.c:628]: Do I2C channels 4 and 5 still need to update the PMER1 register?

File: drivers/i2c/i2c_ite_it8xxx2.c [hunk 5]
```diff
@@ -240,15 +179,15 @@ static void i2c_reset(const struct device *dev)
...
+	IT8XXX2_SMB_SCLKTS(port) = 0;
 	/* Suggested setting of timing registers of 400kHz. */
-	IT83XX_SMB_4P7USL = 0x6;
-	IT83XX_SMB_4P0USL = 0;
-	IT83XX_SMB_300NS = 0x1;
-	IT83XX_SMB_250NS = 0x2;
-	IT83XX_SMB_45P3USL = 0x6a;
-	IT83XX_SMB_45P3USH = 0x1;
-	IT83XX_SMB_4P7A4P0H = 0;
+	IT8XXX2_SMB_4P7USL = 0x3;
```
Comment (keith-zephyr) [inline drivers/i2c/i2c_ite_it8xxx2.c:184]: Why are the timing registers changing?  This should be done in a separate PR.

File: drivers/i2c/i2c_ite_it8xxx2.c [hunk 6]
```diff
@@ -262,61 +201,22 @@ static void i2c_standard_port_set_frequency(const struct device *dev,
 	 * for setting. So we can adjust tlow to meet timing.
 	 * The others use basic 50/100/1000 KHz setting.
 	 */
-	if (freq_khz == 400) {
+	if (freq_khz == I2C_BITRATE_FAST) {
```
Comment (keith-zephyr) [inline drivers/i2c/i2c_ite_it8xxx2.c:204]: The `freq` parameter is now in Hz, not kHz.  Please update.

File: dts/bindings/i2c/ite,enhance-i2c.yaml [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+compatible: "ite,enhance-i2c"
+
+include: i2c-controller.yaml
+
+properties:
+    reg:
+      required: true
+
+    label:
+      required: true
```
Comment (keith-zephyr) [inline dts/bindings/i2c/ite,enhance-i2c.yaml:15]: Label is already required by i2c-controller.yaml.  I don't think it needs to be added here.

**Review by keith-zephyr — CHANGES_REQUESTED** [review]

File: drivers/i2c/i2c_ite_enhance.c (hunk 1 above)
Comment (GTLin08) [reply drivers/i2c/i2c_ite_enhance.c]: Done.

File: drivers/i2c/i2c_ite_enhance.c [hunk 8]
```diff
@@ -0,0 +1,776 @@
...
+	/*
+	 * Let psr(Prescale) = IT83XX_I2C_PSR(p_ch)
+	 * Then, 1 SCL cycle = 2 x (psr + 2) x SMBus clock cycle
+	 * SMBus clock = PLL_CLOCK / clk_div
+	 * SMBus clock cycle = 1 / SMBus clock
+	 * 1 SCL cycle = 1 / (1000 x freq)
+	 * 1 / (1000 x freq) =
+	 *			2 x (psr + 2) x (1 / (PLL_CLOCK / clk_div))
+	 * psr = ((PLL_CLOCK / clk_div) x
+	 *			(1 / (1000 x freq)) x (1 / 2)) - 2
```
[the remark this answers is not in the document]
Comment (GTLin08) [reply drivers/i2c/i2c_ite_enhance.c]: Done.

File: drivers/i2c/i2c_ite_enhance.c (hunk 2 above)
Comment (GTLin08) [reply drivers/i2c/i2c_ite_enhance.c]: Done. Also modified line 726.

File: drivers/i2c/i2c_ite_enhance.c (hunk 3 above)
Comment (GTLin08) [reply drivers/i2c/i2c_ite_enhance.c]: Done.

File: drivers/i2c/i2c_ite_enhance.c (hunk 4 above)
Comment (GTLin08) [reply drivers/i2c/i2c_ite_enhance.c:626]: I2C4 and I2C5 have already set PMER1 register when they are set to FUN3 in pinmux_it8xxx2_set().

File: drivers/i2c/i2c_ite_it8xxx2.c (hunk 5 above)
Comment (GTLin08) [reply drivers/i2c/i2c_ite_it8xxx2.c:184]: OK, separated.

File: drivers/i2c/i2c_ite_it8xxx2.c (hunk 6 above)
Comment (GTLin08) [reply drivers/i2c/i2c_ite_it8xxx2.c]: Done.

File: dts/bindings/i2c/ite,enhance-i2c.yaml (hunk 7 above)
Comment (GTLin08) [reply dts/bindings/i2c/ite,enhance-i2c.yaml]: Removed.

**Review by keith-zephyr — DISMISSED** [review]

**Review by teburd — CHANGES_REQUESTED** [review]

File: drivers/i2c/i2c_ite_enhance.c [hunk 9]
```diff
@@ -0,0 +1,774 @@
...
+ */
+
+#define DT_DRV_COMPAT ite_enhance_i2c
+
+#include <drivers/gpio.h>
+#include <drivers/i2c.h>
+#include <drivers/pinmux.h>
+#include <errno.h>
+#include <logging/log.h>
+		LOG_MODULE_REGISTER(i2c_ite_enhance);
```
[the remark this answers is not in the document]
Comment (GTLin08) [reply drivers/i2c/i2c_ite_enhance.c]: Done.

File: drivers/i2c/i2c_ite_enhance.c [hunk 10]
```diff
@@ -0,0 +1,774 @@
...
+#include "i2c-priv.h"
+#include <soc.h>
+#include <soc_dt.h>
+#include <sys/util.h>
+
+/* Default PLL frequency. */
+#define PLL_CLOCK 48000000
```
[the remark this answers is not in the document]
Comment (GTLin08) [reply drivers/i2c/i2c_ite_enhance.c]: We can get the PLL frequency by reading the IT8XXX2_ECPM_PLLFREQR register. Fixed.

File: drivers/i2c/i2c_ite_enhance.c [hunk 11]
```diff
@@ -0,0 +1,774 @@
...
+	 * SMBus clock cycle = 1 / SMBus clock
+	 * 1 SCL cycle = 1 / freq
+	 * 1 / freq = 2 x (psr + 2) x (1 / (PLL_CLOCK / clk_div))
+	 * psr = ((PLL_CLOCK / clk_div) x (1 / freq) x (1 / 2)) - 2
+	 */
+	if (freq_hz) {
+		/* Get SMBus clock divide value */
+		clk_div = (IT8XXX2_ECPM_SCDCR2 & 0x0F) + 1U;
+		/* Calculate PSR value */
+		psr = (PLL_CLOCK / (clk_div * (2U * freq_hz))) - 2U;
```
[the remark this answers is not in the document]
Comment (GTLin08) [reply drivers/i2c/i2c_ite_enhance.c]: As mentioned above.

File: drivers/i2c/i2c_ite_enhance.c [hunk 12]
```diff
@@ -0,0 +1,774 @@
...
+	/* device does not respond ACK */
+	} else if ((i2c_str & E_HOSTA_BDS_AND_ACK) == E_HOSTA_BDS) {
+		if (IT8XXX2_I2C_CTR(base) & E_ACK)
+			data->err = E_HOSTA_ACK;
+	}
+
+	return data->err;
+}
+
+static void enhanced_i2c_start(const struct device *dev)
```
[the remark this answers is not in the document]
Comment (GTLin08) [reply drivers/i2c/i2c_ite_enhance.c:344]: Fixed. Also modified line 517 and 672.

File: drivers/i2c/i2c_ite_it8xxx2.c [hunk 13]
```diff
@@ -157,53 +125,29 @@ static int i2c_parsing_return_value(const struct device *dev)
...
-			return -EIO;
-	} else {
-		/* The device does not respond ACK */
-		if (data->err == E_HOSTA_ACK)
-			return -ENXIO;
-		else
-			return -EIO;
-	}
+	/* The device does not respond ACK */
+	if (data->err == HOSTA_NACK)
```
Comment (teburd) [inline drivers/i2c/i2c_ite_it8xxx2.c:129]: This function has a few missing enclosing brackets around if branch bodies. https://docs.zephyrproject.org/latest/contribute/index.html#coding-style

**Review by teburd — CHANGES_REQUESTED** [review]
One last code styling issue, looks good otherwise.

File: drivers/i2c/i2c_ite_enhance.c [hunk 14]
```diff
@@ -0,0 +1,803 @@
...
+static void i2c_reset(const struct device *dev)
+{
+	const struct i2c_enhance_config *config = dev->config;
+	uint8_t *base = config->base;
+
+	/* State reset and hardware reset */
+	IT8XXX2_I2C_CTR(base) = E_STS_AND_HW_RST;
+}
+
+static uint32_t i2c_get_pll_frequency(void)
```
Comment (keith-zephyr) [inline drivers/i2c/i2c_ite_enhance.c:205]: This routine is not specific to the I2C module.  Can you move this into [soc.c](https://github.com/zephyrproject-rtos/zephyr/blob/main/soc%2Friscv%2Friscv-ite%2Fit8xxx2%2Fsoc.c).  You can put the prototype into [soc_common.h](https://github.com/zephyrproject-rtos/zephyr/blob/main/soc%2Friscv%2Friscv-ite%2Fcommon%2Fsoc_common.h)

File: drivers/i2c/i2c_ite_it8xxx2.c (hunk 13 above)
Comment (GTLin08) [reply drivers/i2c/i2c_ite_it8xxx2.c]: Fixed. Thank you.

File: drivers/i2c/i2c_ite_enhance.c (hunk 14 above)
Comment (GTLin08) [reply drivers/i2c/i2c_ite_enhance.c]: Fixed. Thank you.

**Review by teburd — DISMISSED** [review]

File: drivers/i2c/Kconfig.it8xxx2 [hunk 15]
```diff
@@ -8,3 +8,10 @@ config I2C_ITE_IT8XXX2
 	  Enable I2C support on it8xxx2_evb.
 	  Supported Speeds: 100kHz, 400kHz and 1MHz.
 	  This driver supports repeated start.
+
+config I2C_ITE_ENHANCE
+	bool "ITE IT8XXX2 I2C enhance driver"
+	depends on SOC_IT8XXX2
```
Comment (gmarull) [inline drivers/i2c/Kconfig.it8xxx2:14]: please default based on dt_compat_status

File: drivers/i2c/i2c_ite_enhance.c [hunk 16]
```diff
@@ -0,0 +1,771 @@
...
+ */
+
+#define DT_DRV_COMPAT ite_enhance_i2c
+
+#include <drivers/gpio.h>
+#include <drivers/i2c.h>
+#include <drivers/pinmux.h>
+#include <errno.h>
+#include <logging/log.h>
+LOG_MODULE_REGISTER(i2c_ite_enhance, CONFIG_I2C_LOG_LEVEL);
```
Comment (gmarull) [inline drivers/i2c/i2c_ite_enhance.c:14]: for consistency, can this be put after include list?

File: drivers/i2c/i2c_ite_enhance.c [hunk 17]
```diff
@@ -0,0 +1,771 @@
...
+#define I2C_LINE_IDLE (I2C_LINE_SCL_HIGH | I2C_LINE_SDA_HIGH)
+
+/*
+ * Structure i2c_alts_cfg is about the alternate function
+ * setting of i2c, this config will be used at initial
+ * time and recover bus.
+ */
+struct i2c_alts_cfg {
+	/* Pinmux control group */
+	const struct device *pinctrls;
```
Comment (gmarull) [inline drivers/i2c/i2c_ite_enhance.c:34]: gentle reminder: you should implement a pinctrl driver.

File: drivers/i2c/i2c_ite_enhance.c [hunk 18]
```diff
@@ -0,0 +1,771 @@
...
+
+	/* Check for NULL pointers */
+	if (dev == NULL) {
+		LOG_ERR("Device handle is NULL");
+		return -EINVAL;
+	}
+	if (msgs == NULL) {
+		LOG_ERR("Device message is NULL");
+		return -EINVAL;
+	}
```
Comment (gmarull) [inline drivers/i2c/i2c_ite_enhance.c:507]: we don't usually make drivers defensive against application programming errors.

File: drivers/i2c/i2c_ite_enhance.c [hunk 19]
```diff
@@ -0,0 +1,771 @@
...
+	case DT_REG_ADDR(DT_NODELABEL(i2c3)):
+		offset = CGC_OFFSET_SMBD;
+		break;
+	case DT_REG_ADDR(DT_NODELABEL(i2c4)):
+		offset = CGC_OFFSET_SMBE;
+		break;
+	case DT_REG_ADDR(DT_NODELABEL(i2c5)):
+		offset = CGC_OFFSET_SMBF;
+		break;
+	}
```
Comment (gmarull) [inline drivers/i2c/i2c_ite_enhance.c:628]: why isn't this stored in each DT instance and retrieved by the driver?

File: drivers/i2c/i2c_ite_enhance.c [hunk 20]
```diff
@@ -0,0 +1,771 @@
...
+
+	/* The pin is set to I2C alternate function of SCL */
+	pinmux_pin_set(config->alts_list[SCL].pinctrls,
+		       config->alts_list[SCL].pin,
+		       config->alts_list[SCL].alt_fun);
+	/* The pin is set to I2C alternate function of SDA */
+	pinmux_pin_set(config->alts_list[SDA].pinctrls,
+		       config->alts_list[SDA].pin,
+		       config->alts_list[SDA].alt_fun);
```
Comment (gmarull) [inline drivers/i2c/i2c_ite_enhance.c:661]: ditto, please be aware of https://github.com/zephyrproject-rtos/zephyr/issues/39740

File: drivers/i2c/i2c_ite_enhance.c [hunk 21]
```diff
@@ -0,0 +1,771 @@
...
+	k_msleep(1);
+
+	/* Set GPIO back to I2C alternate function of SCL */
+	pinmux_pin_set(config->alts_list[SCL].pinctrls,
+		       config->alts_list[SCL].pin,
+		       config->alts_list[SCL].alt_fun);
+	/* Set GPIO back to I2C alternate function of SDA */
+	pinmux_pin_set(config->alts_list[SDA].pinctrls,
+		       config->alts_list[SDA].pin,
+		       config->alts_list[SDA].alt_fun);
```
Comment (gmarull) [inline drivers/i2c/i2c_ite_enhance.c:719]: can this procedure be documented? why sleeps of 1ms?

File: soc/riscv/riscv-ite/it8xxx2/Kconfig.defconfig.series [hunk 22]
```diff
@@ -60,6 +60,10 @@ config I2C_ITE_IT8XXX2
 	default y
 	depends on I2C
 
+config I2C_ITE_ENHANCE
+	default y
+	depends on I2C
```
Comment (gmarull) [inline soc/riscv/riscv-ite/it8xxx2/Kconfig.defconfig.series:65]: please default based on dt_compat_status so that this can be removed.

File: soc/riscv/riscv-ite/it8xxx2/soc.c [hunk 23]
```diff
@@ -51,6 +51,42 @@ static const struct pll_config_t pll_configuration[] = {
...
+		break;
+	case 7:
+		pllfreq = MHZ(96);
+		break;
+	default:
+		return -ERANGE;
+	}
+
+	return pllfreq;
+}
```
Comment (gmarull) [inline soc/riscv/riscv-ite/it8xxx2/soc.c:88]: why isn't this part of a clock control driver?

**Review by gmarull — DISMISSED** [review]

File: drivers/i2c/Kconfig.it8xxx2 (hunk 15 above)
Comment (GTLin08) [reply drivers/i2c/Kconfig.it8xxx2:18]: Done.

File: drivers/i2c/i2c_ite_enhance.c (hunk 16 above)
Comment (GTLin08) [reply drivers/i2c/i2c_ite_enhance.c]: Done.

File: drivers/i2c/i2c_ite_enhance.c (hunk 17 above)
Comment (GTLin08) [reply drivers/i2c/i2c_ite_enhance.c:36]: Pinctrl driver is under development.

File: drivers/i2c/i2c_ite_enhance.c (hunk 18 above)
Comment (GTLin08) [reply drivers/i2c/i2c_ite_enhance.c]: Removed.

File: drivers/i2c/i2c_ite_enhance.c (hunk 19 above)
Comment (GTLin08) [reply drivers/i2c/i2c_ite_enhance.c]: Done.

File: drivers/i2c/i2c_ite_enhance.c (hunk 20 above)
Comment (GTLin08) [reply drivers/i2c/i2c_ite_enhance.c:639]: Ack.

File: drivers/i2c/i2c_ite_enhance.c (hunk 21 above)
Comment (GTLin08) [reply drivers/i2c/i2c_ite_enhance.c]: Add comment.

File: soc/riscv/riscv-ite/it8xxx2/Kconfig.defconfig.series (hunk 22 above)
Comment (GTLin08) [reply soc/riscv/riscv-ite/it8xxx2/Kconfig.defconfig.series]: Removed.

File: soc/riscv/riscv-ite/it8xxx2/soc.c (hunk 23 above)
Comment (GTLin08) [reply soc/riscv/riscv-ite/it8xxx2/soc.c:88]: The clock control driver will be implemented later.

File: dts/bindings/i2c/ite,it8xxx2-i2c.yaml [hunk 24]
```diff
@@ -33,3 +30,10 @@ properties:
       description: Configuration of I2C SCL and SDA pinmux controller.
                    The SCL pin must be specified first and the SDA pin
                    second in the pinctrl-0 array.
+
+    clock-gate-offset:
+      type: int
+      required: true
+      description: |
+        The clock gate offsets combine the register offset from
+        ECPM_BASE and the mask within that register into one value.
```
Comment (mbolivar-nordic) [inline dts/bindings/i2c/ite,it8xxx2-i2c.yaml:39]: It seems like the following properties are common to both I2C controllers:

- reg
- interrupts
- port-num
- gpio-dev
- pinctrl-0
- clock-gate-offset

Only prescale-scl-low is unique to the 'enhanced' driver. So for maintainability reasons you may want to consider using `include` as is done here for example:

https://github.com/zephyrproject-rtos/zephyr/blob/main/dts/bindings/i2c/nordic%2Cnrf-twi-common.yaml
https://github.com/zephyrproject-rtos/zephyr/blob/main/dts/bindings/i2c/nordic%2Cnrf-twi.yaml
https://github.com/zephyrproject-rtos/zephyr/blob/main/dts/bindings/i2c/nordic%2Cnrf-twim.yaml

This is not a blocking request, though, just something for you to consider. Thanks for the patch!

**Review by mbolivar-nordic — DISMISSED** [review]
Approval is for the DT portions only; I didn't review the rest

File: dts/bindings/i2c/ite,it8xxx2-i2c.yaml (hunk 24 above)
Comment (GTLin08) [reply dts/bindings/i2c/ite,it8xxx2-i2c.yaml]: Yes, It can be better maintained, thanks for your advise. Done.

**Comment by GTLin08** [comment]
Dear @gmarull, 
Could you review the requested changes, thanks!

**Comment by GTLin08** [comment]
Hi @teburd, @keith-zephyr, @mbolivar-nordic 
I need to get your approval again, thanks!

**Review by teburd — APPROVED** [review]

**Review by keith-zephyr — APPROVED** [review]

**Review by brockus-zephyr — APPROVED** [review]

**Comment by GTLin08** [comment]
Dear sirs,

Truly grateful for helps. However, we’d like to invite you back to review the PR #42812 which is about the I2C 50KHz option. Thanks again for your great patience and guidance.
